docs/rfd.md

```diff
@@ -0,0 +1,478 @@
+
+## Scope
+
+Due to the nature of this project, the following limitations will be made:
+
+1. Only linux/amd64 architecture will be supported.
+2. Only Cgroups v2 will be supported (however, sufficient abstraction will be put in place to allow implenting v1 support if desired).
+3. The server must be run with root privileges.
+4. Authorization will be limited in scope and will not impose restrictions on commands that can be run. The authorization model is limited to restricting access to specific API methods.
+5. Test coverage will be limited to a few packages only.
+6. Authentication will be limited to mTLS only, and will use pre-generated certificates.
+7. Everything is stored in memory, and no persistent storage will be used. See [Security Considerations](#security-considerations) for more details.
+8. Only a small subset of the possible ways to limit resources will be implemented:
+
+   - cpu limits using cpu.max only, period hard coded to the default 100000, no weight/shares support, no cpuset support
+   - memory limits using memory.high and memory.max only, no swap or hugepages support
+   - device io limits using io.max only, no weight or latency support
+
+## Disclaimer
+
+This project is **not** intended for real-world use, and does not implement many of the the features that would be necessary to make it safe to use in a production environment.
+
+## Details
+
+The three main components of the job server are:
+
+1. A server that provides a GRPC API to allow clients to start and stop jobs, list and query the status of existing jobs, and stream logs of running or completed jobs.
+2. A set of generic and reusable library packages that are used to implement the server and API.
+3. A user-friendly CLI client that allows users to interact with the job server.
+
+### Proto Specification
+
+The Job Server API is composed of a single service with five methods, defined as follows:
+
+```protobuf
+syntax = "proto3";
+
+package job.v1;
+
+import "google/protobuf/empty.proto";
+import "google/protobuf/timestamp.proto";
+
+option go_package = "github.com/kralicky/jobserver/pkg/apis/job/v1";
+
+service Job {
+  // Starts a new job, and returns its id if it was started successfully.
+  //
+  // The job will be run asynchronously; this method does not wait for it
+  // to complete.
+  //
+  // If resource limits are provided, the job will be run in its own cgroup
+  // with all the given limits applied. Any combination of the allowed limit
+  // types may be provided; it is not necessary to set all of them.
+  rpc Start(JobSpec) returns (JobId);
+
+  // Stops a running job. Once stopped, this method will wait until the job has
+  // completed before returning.
+  //
+  // This will first attempt to stop the job's process using SIGTERM, but if the
+  // process does not exit within a short grace period, it will be forcefully
+  // killed with SIGKILL.
```

Comment:
Leave this in if you feel like it'd be an interesting addition code-wise, otherwise just kill the process immediately.

Reply:
I had this in here to account for being able to run jobs that run forever until you send SIGINT or SIGTERM (tail -f, journalctl -f, etc). But it might end up being more trouble than it's worth, so I will keep that in mind.